### PR TITLE
Add conditional dependency on LLVM packages for Mono

### PR DESCRIPTION
**Pull request by directhex**
These are only restored or used when /p:MonoEnableLLVM=true

Creating as draft for now - there's one breakage I want help with (the restore only copies LLVM to the staging tree the second time it's run)

### CODE REVIEW

**Comment by directhex** [comment]
Oh, forgot to add it to CI, too. I'll do that...

File: eng/Version.Details.xml [hunk 1]
```diff
@@ -150,5 +150,29 @@
...
+    </Dependency>
+    <Dependency Name="runtime.win-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk" Version="6.0.1-alpha.1.20073.2">
+      <Uri>https://github.com/dotnet/llvm-project</Uri>
+      <Sha>320e484510d00c6112cfe972e3758fdcaa955864</Sha>
+    </Dependency>
+    <Dependency Name="runtime.win-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools" Version="6.0.1-alpha.1.20073.2">
+      <Uri>https://github.com/dotnet/llvm-project</Uri>
+      <Sha>320e484510d00c6112cfe972e3758fdcaa955864</Sha>
+    </Dependency>
+    <Dependency Name="runtime.osx.10.12-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk" Version="6.0.1-alpha.1.20073.2">
```
Comment (marek-safar) [inline eng/Version.Details.xml]: Why does osx have to have the version in?

Comment (directhex) [reply eng/Version.Details.xml]: I have no idea. I copy-pasted that logic from Mono into LLVM: https://github.com/dotnet/runtime/blob/master/src/mono/netcore/nuget/Directory.Build.props#L17 -> https://github.com/dotnet/llvm-project/blob/release/6.x/nuget/Directory.Build.props#L14

File: eng/Version.Details.xml [hunk 2]
```diff
@@ -150,5 +150,29 @@
       <Uri>https://github.com/dotnet/runtime-assets</Uri>
       <Sha>4e8d82a2bdc31db2a3d392b4021b0c9ab19a4228</Sha>
     </Dependency>
+    <Dependency Name="runtime.linux-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk" Version="6.0.1-alpha.1.20073.2">
```
Comment (akoeplinger) [inline eng/Version.Details.xml]: I don't think we need to list all platforms/architectures here, the `runtime.native.System.IO.Ports` on line 19 doesn't do it either. We might need to change how we're producing that package.

Comment (directhex) [reply eng/Version.Details.xml]: `runtime.native.System.IO.Ports` is a metapackage, and we don't have a metapackage for LLVM. We can add one, sure

Comment (akoeplinger) [reply eng/Version.Details.xml]: Yeah I think we should do that.

Comment (marek-safar) [reply eng/Version.Details.xml]: We should as the same package will be used for AOT tooling packages

Comment (directhex) [reply eng/Version.Details.xml]: We're going to use one metapackage to pull in AOT support (i.e. `llc` and `opt` binaries) for every single supported architecture in one go?

Comment (marek-safar) [reply eng/Version.Details.xml]: No, it will be RID driven for host (and possibly target) but the package should be identical to our internal dependency

Comment (directhex) [reply eng/Version.Details.xml]: That's the split already in place here - `runtime.linux-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk` contains LLVM for build-time (i.e. what we need internally to compile LLVM JIT Mono), `runtime.linux-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools` contains `opt` and `llc`, the two binaries needed for LLVM AOT support (i.e. end users would receive this, one way or another). If we want to rename things, I'm open to suggestions.

**Comment by directhex** [comment]
Okay. We're green, and building LLVM Mono. I'm going to try and get LLVM metapackages generated today, for cleaner dependency handling, but will mark this PR as not-draft tomorrow regardless of status (the metapackage is a nice-to-have, but explicitly not a blocker on functional LLVM builds)

File: src/mono/mono.proj [hunk 3]
```diff
@@ -19,7 +19,7 @@
         <_MonoExtraCXXFLAGS>-O2 -g</_MonoExtraCXXFLAGS>
     </PropertyGroup>
     <PropertyGroup>
-        <_MonoConfigureParams Condition="$(MonoEnableLLVM) == true">$(_MonoConfigureParams) --enable--llvm </_MonoConfigureParams>
+        <_MonoConfigureParams Condition="$(MonoEnableLLVM) == true">$(_MonoConfigureParams) --enable--llvm --with-llvm=$(MonoObjDir)llvm </_MonoConfigureParams>
```
Comment (akoeplinger) [inline src/mono/mono.proj]:
```suggestion
        <_MonoConfigureParams Condition="'$(MonoEnableLLVM)' == 'true'">$(_MonoConfigureParams) --enable--llvm --with-llvm=$(MonoObjDir)llvm </_MonoConfigureParams>
```

File: src/mono/mono.proj [hunk 4]
```diff
@@ -73,6 +73,7 @@
 
   <Target Name="Restore">
     <MSBuild Projects="$(MonoProjectRoot)netcore\nuget\packages.builds" Targets="Restore" />
+    <MSBuild Condition=" '$(MonoEnableLLVM)' == true " Projects="$(MonoProjectRoot)llvm\llvm-init.proj" Targets="Restore" Properties="MonoObjDir=$(MonoObjDir)" />
```
Comment (akoeplinger) [inline src/mono/mono.proj]:
```suggestion
    <MSBuild Condition="'$(MonoEnableLLVM)' == 'true'" Projects="$(MonoProjectRoot)llvm\llvm-init.proj" Targets="Restore" Properties="MonoObjDir=$(MonoObjDir)" />
```

File: src/mono/llvm/llvm-init.proj [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    <PackageReference Condition="'$(OSGroup)' == 'OSX'" Include="runtime.osx.10.12-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools" Version="$(runtimeosx1012x64MicrosoftNETCoreRuntimeMonoLLVMToolsVersion)" GeneratePathProperty="True" />
+  </ItemGroup>
+
+  <Target Name="CopyLLVMToTree" AfterTargets="Restore">
+    <ItemGroup>
+      <LLVMFiles Include="$(NuGetPackageRoot)\runtime.$(NETCoreSdkRuntimeIdentifier).microsoft.netcore.runtime.mono.llvm.sdk\$(runtimelinuxx64MicrosoftNETCoreRuntimeMonoLLVMSdkVersion)\tools\$(NETCoreSdkRuntimeIdentifier)\**" />
+      <LLVMFiles Include="$(NuGetPackageRoot)\runtime.$(NETCoreSdkRuntimeIdentifier).microsoft.netcore.runtime.mono.llvm.tools\$(runtimelinuxx64MicrosoftNETCoreRuntimeMonoLLVMToolsVersion)\tools\$(NETCoreSdkRuntimeIdentifier)\**" />
+    </ItemGroup>
+
+    <Copy SourceFiles="@(LLVMFiles)" DestinationFolder="$(MonoObjDir)\llvm\%(RecursiveDir)">
```
Comment (akoeplinger) [inline src/mono/llvm/llvm-init.proj]: we should create a property in mono.proj for `$(MonoObjDir)\llvm` i.e. something like `$(MonoLLVMDir)` and pass that down.

File: src/mono/llvm/llvm-init.proj [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    <PackageReference Condition="'$(OSGroup)' == 'OSX'" Include="runtime.osx.10.12-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk" Version="$(runtimeosx1012x64MicrosoftNETCoreRuntimeMonoLLVMSdkVersion)" GeneratePathProperty="True" />
+    <PackageReference Condition="'$(OSGroup)' == 'OSX'" Include="runtime.osx.10.12-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools" Version="$(runtimeosx1012x64MicrosoftNETCoreRuntimeMonoLLVMToolsVersion)" GeneratePathProperty="True" />
+  </ItemGroup>
+
+  <Target Name="CopyLLVMToTree" AfterTargets="Restore">
+    <ItemGroup>
+      <LLVMFiles Include="$(NuGetPackageRoot)\runtime.$(NETCoreSdkRuntimeIdentifier).microsoft.netcore.runtime.mono.llvm.sdk\$(runtimelinuxx64MicrosoftNETCoreRuntimeMonoLLVMSdkVersion)\tools\$(NETCoreSdkRuntimeIdentifier)\**" />
+      <LLVMFiles Include="$(NuGetPackageRoot)\runtime.$(NETCoreSdkRuntimeIdentifier).microsoft.netcore.runtime.mono.llvm.tools\$(runtimelinuxx64MicrosoftNETCoreRuntimeMonoLLVMToolsVersion)\tools\$(NETCoreSdkRuntimeIdentifier)\**" />
```
Comment (akoeplinger) [inline src/mono/llvm/llvm-init.proj]: I assume the `GeneratePathProperty` doesn't work given you're manually constructing the path here? we should remove the property on the PackageReference then.

Comment (directhex) [reply src/mono/llvm/llvm-init.proj]: GeneratePathProperty creates the property in a g.props file, but doesn't add it to the current task execution, so you essentially cannot access it during a `Restore` task or any tasks executed via AfterTargets.

So no, turns out it's useless.

**Comment by directhex** [comment]
Those aren't my failures, they're the usual infra breakage.

I'm trying & failing to figure out how to drive the metapackage ("lineup package") creation functionality in Arcade.

File: eng/pipelines/mono/templates/build-job.yml [hunk 7]
```diff
@@ -21,18 +22,27 @@ jobs:
...
+      displayName: ${{ format('Mono LLVM Product Build {0}{1} {2} {3}', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+
 
     # Run all steps in the container.
     # Note that the containers are defined in platform-matrix.yml
     container: ${{ parameters.container }}
 
     timeoutInMinutes: ${{ parameters.timeoutInMinutes }}
 
+
```
Comment (akoeplinger) [inline eng/pipelines/mono/templates/build-job.yml]: additional newline

**Review by akoeplinger — APPROVED** [review]

**Comment by directhex** [comment]
The hell are some of these failures? I haven't touched CoreCLR Windows

File: eng/pipelines/runtime.yml [hunk 8]
```diff
@@ -250,6 +250,62 @@ jobs:
           eq(dependencies.checkout.outputs['SetPathVars_mono.containsChange'], true),
           eq(variables['isFullMatrix'], true))
 
+#
+# Build Mono LLVM debug
+# Only when libraries or mono changed
```
Comment (marek-safar) [inline eng/pipelines/runtime.yml]: Shouldn't this also build when llvm dependency is updated?

Comment (directhex) [reply eng/pipelines/runtime.yml]: I... have no idea how to represent that.

Comment (akoeplinger) [reply eng/pipelines/runtime.yml]: @ViktorHofer can you help?

**Comment by directhex** [comment]
I can't get Helix to give me the failure log from that run, I just get blob not found errors

File: eng/pipelines/runtime.yml (hunk 8 above)
Comment (ViktorHofer) [reply eng/pipelines/runtime.yml]: It's already doing that. Everything's included by default + the additional exclude minus the exclude: https://github.com/dotnet/runtime/blob/master/eng/pipelines/runtime.yml#L74-L89.

**Comment by directhex** [comment]
OK well I guess the failures I was getting were flakes

**Review by akoeplinger — APPROVED** [review]

**Review by EgorBo — APPROVED** [review]

**Pull request merged**